### PR TITLE
[FIXED JENKINS-35160] - Job deletion: Wait up to 15 seconds for interrupted builds to complete

### PR DESCRIPTION
**Pull request by stephenc**
- Also now aware of concurrent builds

See [JENKINS-35160](https://issues.jenkins-ci.org/browse/JENKINS-35160)

@reviewbybees

### CODE REVIEW

File: core/src/main/java/hudson/model/Job.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Job.java]: It won't work for the Promoted Builds and other such... exotic implementations

File: core/src/main/java/hudson/model/Job.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Job.java]: Lacks logging

File: core/src/main/java/hudson/model/Job.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Job.java]: Lacks logging

File: core/src/main/java/hudson/model/Job.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Job.java]: Should it also go to the system log?

File: core/src/main/java/hudson/model/Job.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Job.java]: It should be `isActive()`. E.g. if the executor is starting up/shutting down, we should not ignore the task check 🐛

File: core/src/main/java/hudson/model/Job.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Job.java]: 🐛 this method uses cached list of builds. If the new build gets started due to whatever reason, during the deletion cycle, these builds won't be noticed.

File: core/src/main/java/hudson/model/Job.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Job.java]: Risk of false negatives since this is a value from the past

File: core/src/main/java/hudson/model/Job.java
Comment (stephenc) [inline core/src/main/java/hudson/model/Job.java]: need to not hold the lock while interrupting or the interrupted threads may be unable to complete

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
🐛 for the concurrency flaws. Ideally the job should be marked for deletion (e.g. boolean flag), and after setting this flag the job should prevent any attempts to schedule it to the queue. And maybe the flag should be released if the deletion fails (or not?).

Another 🐛  for using `isAlive()` for the executor

File: core/src/main/java/hudson/model/Job.java
Comment (jglick) [inline core/src/main/java/hudson/model/Job.java]: Better to factor into a helper method. Better yet, introduce `Computer.getAllExecutors()`, since that is a common problem. (See `Executor.of` etc.)

File: core/src/main/java/hudson/model/Job.java
Comment (jglick) [inline core/src/main/java/hudson/model/Job.java]: Maybe easier to just check for

```
Queue.Executable exec = e.getCurrentExecutable();
if (exec instanceof Run && ((Run) exec).getParent() == this) {
```

File: core/src/main/java/hudson/model/Job.java
Comment (jglick) [inline core/src/main/java/hudson/model/Job.java]: And then we could use simply `Run.isLogUpdated()`.

File: core/src/main/java/hudson/model/Job.java
Comment (jglick) [inline core/src/main/java/hudson/model/Job.java]: Better consider that architecturally deprecated.

**Review by jglick — DISMISSED** [review]

File: core/src/main/java/hudson/model/Job.java
Comment (jglick) [inline core/src/main/java/hudson/model/Job.java]: I do not see why. The user has gotten a failure message (I hope—test it?); the job has not been deleted; there is nothing for the admin to do.

**Comment by stephenc** [comment]
@oleg-nenashev if the executor is interrupted the thread will die: https://github.com/jenkinsci/jenkins/commit/111a8d364c4a7a1afb8c8a88a7ab6809bb2a19dc#diff-4da8c72b731d029f61f01ed41568d209R248 so the only way to know that the executor has completed is to observe  `isAlive()`

Other issues addressed

**Comment by oleg-nenashev** [comment]
@stephenc maybe it also addresses https://issues.jenkins-ci.org/browse/JENKINS-32783

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/AbstractItem.java]: Maybe it worth putting this warning into UI as well (e.g. in the summary page near the "disabled" status)

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/AbstractItem.java]: Maybe they should, but not sure. It makes FolderDelition a separate long background task. Which is probably fine, but out of the scope of this PR imho

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/AbstractItem.java]: 🐛  No `CauseOfInterruption`, which would be really useful in this code. The default behavior will be an emty cause in such case: https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/model/Executor.java#L190

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/AbstractItem.java]: Same, missing cause

**Review by oleg-nenashev — DISMISSED** [review]
I have a minor :ant: regarding the missing Interruption cause, but the most of the code looks good to me. IMHO some testing like ATH and PCT would not hurt in such case due to the fix complexity and potential risk of severe regressions.

Also dismissing the review from @jglick since the code has changed significantly

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (stephenc) [inline core/src/main/java/hudson/model/AbstractItem.java]: 1. The job is going to be deleted
2. As the this is the user who will be triggering the interrupt the `UserInterruption` cause will be injected

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (stephenc) [inline core/src/main/java/hudson/model/AbstractItem.java]: Nope: https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/model/Executor.java#L183-L196

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (stephenc) [inline core/src/main/java/hudson/model/AbstractItem.java]: We are recursively interrupting all jobs in one go here, any deletion will be at most 15 seconds

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (stephenc) [inline core/src/main/java/hudson/model/AbstractItem.java]: For all of 15 seconds!!!

**Comment by stephenc** [comment]
@oleg-nenashev I argue that your bug was incorrect, the interruption is a user triggered action and will have a cause reflecting that

File: core/src/main/java/hudson/model/Executor.java
Comment (stephenc) [inline core/src/main/java/hudson/model/Executor.java]: I think you are arguing for more complex code on the grounds of performance in the absence of a proven performance degradation. That smells like premature optimization

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/AbstractItem.java]: okay

Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/AbstractItem.java]: Still may be a case if the timeout gets configurable

Comment (stephenc) [inline core/src/main/java/hudson/model/AbstractItem.java]: Then whoever adds that feature can add the UI indicator. Solved

File: core/src/main/java/hudson/model/Executor.java
Comment (stephenc) [inline core/src/main/java/hudson/model/Executor.java]: if you have lots of one-off executors then the majority of cases will be checking the one-off executors, so we save two ArrayList and two Iterator allocations, plus escape analysis can reveal that the one list is only ever used for iteration and the method can be inlined as it is small... the JVM is smarter than you think. (maybe not that smart, but there is hope!)

**Comment by stephenc** [comment]
@oleg-nenashev @jglick ping also @jenkinsci/code-reviewers

**Comment by oleg-nenashev** [comment]
@stephenc removed the bug according to explanation

**Comment by oleg-nenashev** [comment]
@reviewbybees any feedback?

**Comment by stephenc** [comment]
@oleg-nenashev per policy, 7 days so @reviewbybees done

File: core/src/main/java/hudson/model/Computer.java
Comment (stephenc) [inline core/src/main/java/hudson/model/Computer.java]: Why do you want to add another layer of indirection, plus we have given you a copy so you can do what you want with it

File: core/src/main/java/jenkins/model/queue/ItemDeletion.java
Comment (stephenc) [inline core/src/main/java/jenkins/model/queue/ItemDeletion.java]: No because plugins may need access to perform additional checks and prevent work on items that are registered

File: core/src/main/java/hudson/model/Computer.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/Computer.java]: Because that's how I interpret the "read-only" part of the Javadoc.

File: core/src/main/java/jenkins/model/queue/ItemDeletion.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/queue/ItemDeletion.java]: Will be likely non-backportable then

**Review by oleg-nenashev — APPROVED** [review]
I agree with the fix since it definitely does not make the behavior worse than it was before. I am concerned about adding another non-configurable magic number for timeout, but I accept delaying it.

I will be likely against backporting the fix into .2 due to the potential impact of the changes. Also non-Restricted API

File: core/src/main/java/hudson/model/Computer.java
Comment (rsandell) [inline core/src/main/java/hudson/model/Computer.java]: Yea, `read-only` should be removed from the javadoc, `snapshot view` should be enough, since it actually isn't read-only :)

File: core/src/main/java/jenkins/model/queue/ItemDeletion.java
Comment (rsandell) [inline core/src/main/java/jenkins/model/queue/ItemDeletion.java]: It could be backported, but restricted in the backport

**Comment by oleg-nenashev** [comment]
On hold for now since we are waiting for the 2.54 release. Will merge afterwards

File: test/src/test/java/hudson/model/JobTest.java
Comment (rsandell) [inline test/src/test/java/hudson/model/JobTest.java]: Might result in flaky builds on slower build agents.

**Comment by rsandell** [comment]
:bee:

**Comment by oleg-nenashev** [comment]
@reviewbybees done
No merge right now pls

**Comment by oleg-nenashev** [comment]
@jenkinsci/code-reviewers we would appreciate some additional feedback. It **may** potentially break some use-cases with deletion of projects with pending builds, though these use-cases were not working reliably anyway. If you are fine, my plan is to merge it towards the next weekly on Friday.

File: core/src/main/java/hudson/model/Computer.java
Comment (jglick) [inline core/src/main/java/hudson/model/Computer.java]: Would prefer to just use `unmodifiableList`.

File: core/src/main/java/hudson/model/queue/Tasks.java
Comment (jglick) [inline core/src/main/java/hudson/model/queue/Tasks.java]: Can do that now.

File: core/src/main/java/jenkins/model/queue/ItemDeletion.java
Comment (jglick) [inline core/src/main/java/jenkins/model/queue/ItemDeletion.java]: Should be `@Restricted` unless there is a tested & documented use case for accessing it from a plugin (which I doubt—a real API would be designed differently).

File: core/src/main/java/jenkins/model/queue/ItemDeletion.java
Comment (jglick) [inline core/src/main/java/jenkins/model/queue/ItemDeletion.java]: Pity that `Lock` is not `AutoCloseable`.

**Review by jglick — APPROVED** [review]
Did not follow all of the code here but overall looks right.

**Comment by daniel-beck** [comment]
> Should be `@Restricted` unless there is a tested & documented use case for accessing it from a plugin (which I doubt—a real API would be designed differently).

:frowning_face:

**Comment by oleg-nenashev** [comment]
@jglick @daniel-beck 

> Should be @Restricted unless there is a tested & documented use case for accessing it from a plugin (which I doubt—a real API would be designed differently).

Makes sense to Restrict in a follow-up PR.

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (jglick) [inline core/src/main/java/hudson/model/AbstractItem.java]: Unusable API; see https://github.com/jenkinsci/cloudbees-folder-plugin/pull/112.

**Pull request merged**
